[AST] Enforce that TupleExprs and ParenExprs dont have param flags

hamishknight opened the pull request:
Tuple and paren types may still have parameter type flags in certain cases, but they should never make it into expressions.

This requires formalizing the rules around whether we can convert a function with N parameters into a function with a single N-element tuple parameter. Previously, only functions with variadic, autoclosure, or inout parameter flags were rejected. Other parameter flags were preserved on the resulting tuple, which could cause issues in downstream code, as tuples shouldn't have parameter flags. This patch ensures we either reject the parameter flags, or strip them from the tuple type.

By default, a parameter with non-empty parameter flags will now be rejected, with an exception for ownership flags, `@_nonEphemeral`, and `@noDerivative`. 

This means that tupling `isolated` parameters is now forbidden. An additional rule has also been added that forbids tupling into a `@differentiable` function.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2218,9 +2217,11 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
     auto input = AnyFunctionType::composeTuple(getASTContext(), params,
-                                               /*canonicalVararg=*/false);
-
+                                               /*wantParamFlags*/ false);
```
Wouldn't canImplodeParams() return false in this case?

hamishknight replied:
It only currently returns false for vargs, inout, and autoclosure, but there may still be other parameter flags like `__owned` or `@_nonEphemeral` to strip.

xedin replied:
I guess the question is whether `canImplodeParams` should return false if any of the parameters here have flags? I think the answer is yes if it's only valid to convert "simple" parameters lists into tuples otherwise there is a risk that some new parameter flag would produce a valid solution but then crash in runtime?

hamishknight replied:
That's a fair point, I *believe* we should support stripping for ownership flags such as `__owned`, as we can handle the function conversion through a thunk, and we have test cases for that in the test suite. I believe we should also support stripping for `@_nonEphemeral`, as the attribute can be freely stripped off a function type anyways. Talking with @rxwei, it seems we should also support stripping `@noDerivative` (as long as the destination isn't `@differentiable`).

However I do agree that defaulting to returning false from `canImplodeParams` for any other parameter flag makes sense. For `isolated` in particular that seems like the right behavior.

I've gone ahead and updated the PR to default to returning false, but with some exceptions for the above mentioned cases.

hamishknight commented:
@swift-ci please test source compatibility Release

hamishknight committed:
[AST] Simplify AnyFunctionType::composeTuple

Remove the canonicalVararg parameter and
CanParamArrayRef wrapper. Almost none of the
callers want canonicalVararg, and the one that
does calls `getCanonicalType` on the result
anyway.

hamishknight committed:
[CS] Formalize param flag handling for imploding params

By default avoid imploding params that have parameter
flags, but carve out exceptions for ownership flags,
which can be thunked, and `@_nonEphemeral` which can
be freely dropped without issue.

hamishknight committed:
[AST] Enforce that TupleExprs and ParenExprs dont have param flags

Tuple and paren types may still have parameter
type flags in certain cases, but they should never
make it into expressions.

hamishknight committed:
[AutoDiff] [CS] Formalize function tupling behavior for @differentiable

Reject tupling into a `@differentiable` function,
and allow the stripping of `@noDerivative` given
we're also losing `@differentiable`.

hamishknight commented on test/Concurrency/isolated_parameters.swift:
```diff
@@ -102,3 +102,13 @@ actor TestActor {
 func redecl(_: TestActor) { } // expected-note{{'redecl' previously declared here}}
 func redecl(_: isolated TestActor) { } // expected-error{{invalid redeclaration of 'redecl'}}
+
+func tuplify<Ts>(_ fn: (Ts) -> Void) {} // expected-note {{in call to function 'tuplify'}}
+
+@available(SwiftStdlib 5.5, *)
+func testTuplingIsolated(_ a: isolated A, _ b: isolated A) {
+  // FIXME: We shouldn't duplicate the "cannot convert value of type" diagnostic (SR-15179)
+  tuplify(testTuplingIsolated)
+  // expected-error@-1 {{generic parameter 'Ts' could not be inferred}}
+  // expected-error@-2 2{{cannot convert value of type '(isolated A, isolated A) -> ()' to expected argument type '(Ts) -> Void'}}
+}
```
@DougGregor Mind double checking that we indeed shouldn't support tupling for `isolated` parameters?

DougGregor replied:
We shouldn't support this, since `isolated` parameters are new.

hamishknight commented:
@rxwei Mind taking a look at 77e6c083f3ad5023e43729b1369d2db853480a5f?

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin reviewed (approved):
LGTM!

hamishknight commented:
@swift-ci please test Linux
